---
title: "PKR CRISPR screen"
author: "Leonard Blaschek, PostDoc"
output: 
  html_document:
    theme: yeti
    highlight: kate
    toc: true
    toc_float: true
    toc_depth: 2
    includes:
     after_body: PKR_network.html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  warning = FALSE, message = FALSE
)
library(ontologyIndex)
library(showtext)
library(tidyverse)
library(ggbeeswarm)
library(ggtext)
library(gt)
library(patchwork)
library(tukeygrps)
library(seqinr)
# library(biomartr)
# library(metablastr)
# library(reticulate)
library(kableExtra)
# if (dir.exists("/data/")) {
#   use_python("/home/leonard/Applications/miniconda3/envs/blast/bin/python") # office
# } else {
#   use_python("/home/leonard/Applications/mambaforge/envs/blast/bin/python") # laptop
# }
# reticulate::py_config()
options(timeout = 30000)
options(knitr.kable.NA = "")
showtext_opts(dpi = 300)

## import Helvetica ####
# font_add(
#   "Helvetica",
#   regular = "/prop_fonts/01. Helvetica   [1957 - Max Miedinger]/HelveticaNeueLTStd-Lt.otf",
#   italic = "/prop_fonts/01. Helvetica   [1957 - Max Miedinger]/HelveticaNeueLTStd-LtIt.otf",
#   bold = "/prop_fonts/01. Helvetica   [1957 - Max Miedinger]/HelveticaNeueLTStd-Bd.otf",
#   bolditalic = "/prop_fonts/01. Helvetica   [1957 - Max Miedinger]/HelveticaNeueLTStd-BdIt.otf"
# )
font_add(
  "Helvetica",
  regular = "/prop_fonts/01. Helvetica   [1957 - Max Miedinger]/HelveticaLTStd-Light.otf",
  italic = "/prop_fonts/01. Helvetica   [1957 - Max Miedinger]/HelveticaLTStd-LightObl.otf",
  bold = "/prop_fonts/01. Helvetica   [1957 - Max Miedinger]/HelveticaLTStd-Bold.otf",
  bolditalic = "/prop_fonts/01. Helvetica   [1957 - Max Miedinger]/HelveticaLTStd-BoldObl.otf"
)
font_add(
  "Univers",
  regular = "/usr/share/fonts/propietary_fonts/Univers/UniversLTStd.otf"
  )
showtext_auto()

#### generating plot theme ####
font_add(
  "Futura",
  regular = "FuturaStd-Medium.otf",
  italic = "FuturaStd-MediumOblique.otf",
  bold = "FuturaStd-Bold.otf",
  bolditalic = "FuturaStd-BoldOblique.otf"
)
showtext_auto()

pal_ostwald_disc <- c(
  "#275d95",
  "#e8c245",
  "#d25952",
  "#7ab6c4"
)

blues <- c(
  pal_ostwald_disc[1],
  colorspace::lighten(pal_ostwald_disc[1], 0.4),
  colorspace::lighten(pal_ostwald_disc[1], 0.7)
)

text_size <- 9
ggtext_size <- text_size / (14 / 5)

#### generating plot theme ####
theme_leo <- function(base_size = text_size,
                      base_family = "Futura") {
  theme_minimal(
    base_size = base_size,
    base_family = base_family
  ) %+replace%
    theme(
      strip.text = element_text(hjust = 0, face = "italic"),
      axis.ticks = element_blank(),
      axis.text.x = element_text(
        colour = "black",
        margin = margin(1, 1, 1, 1),
        size = text_size
      ),
      axis.text.y = element_text(
        colour = "black",
        angle = 0,
        vjust = 0.5,
        hjust = 1,
        margin = margin(1, 1, 1, 1),
        size = text_size
      ),
      # axis.title = element_blank(),
      plot.title = element_text(size = text_size, hjust = 0),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.spacing = unit(1.5, "mm"),
      legend.position = "bottom",
      legend.text = element_text(size = rel(0.8)),
      legend.key.height = unit(4, "mm"),
      complete = TRUE
    )
}

# convenience figure size functions
ggtext_size <- 8 / (14 / 5)
cm_size <- function(x) x / 2.54
twocol <- 18 / 2.54
onehalfcol <- 14 / 2.54
onecol <- 9 / 2.54

pal_ostwald_disc <- c(
  "#275d95",
  "#e8c245",
  "#d25952"
)

pal_ostwald_disc_long <- c(
  "#8fab1d",
  "#2d7d73",
  "#1d566f",
  "#275d95",
  "#e8c245",
  "#d25952"
)

#### machine dependent paths ####
datapath <- ifelse(dir.exists("/data/2022_CRISPR_screen/"), "/data/", "/run/media/leonard/CPH/data/")

out_path <- paste0(datapath, "2022_CRISPR_screen/phenotyping/")
```

## PKR library

```{r}
lib <- read_tsv(paste0(datapath, "2022_CRISPR_screen/Kinases family CRISPR library.tsv"))

targets <- lib |> 
  mutate(targets = str_extract_all(targeted_genes, "AT[:alnum:]+")) |> 
  pull(targets) |> 
  unlist()

unique_genes <- unique(targets)

print(paste(nrow(lib), "sgRNAs targeting", length(unique_genes), "unique genes encoding for receptors, kinases, phosphatases or receptor ligands."))

hits <- tibble("gene" = unique_genes)

data_path <- "/data/PUP-IT/network/"

obo <- get_ontology(paste0(data_path, "databases/go.obo"), extract_tags = "everything")

annotated_terms <- obo$name |> 
  as_tibble(rownames = "GO_ID") |> 
  dplyr::rename("name" = "value")
namespace_terms <- obo$namespace |> 
  unlist() |> 
  as_tibble(rownames = "GO_ID") |> 
  dplyr::rename("namespace" = "value")

annotation <- full_join(annotated_terms, namespace_terms)

gff <- ape::read.gff(paste0(data_path, "databases/Araport11_GFF3_genes_transposons.current.gff")) |> 
  mutate(gene = str_extract(attributes, "ID=([:alnum:]+)", group = 1)) |> 
  filter(gene != "ath") |> # remove miRNAs
  mutate("Description" = str_extract(attributes, "description=([^;]+)", group = 1)) |> 
  select(gene, Description) |> 
  group_by(gene) |> 
  arrange(Description) |> 
  slice_head(n = 1)

terms <- read_tsv(
  paste0(data_path, "databases/tair.gaf"),
  comment = "!",
  col_names = c(
    "DB",
    "DB_ID",
    "DB_Symbol",
    "Relation",
    "GO_ID",
    "DB_References",
    "Evidence",
    "With",
    "Aspect",
    "Name",
    "Synonym",
    "Type",
    "Taxon",
    "Date",
    "By",
    "Extension",
    "Form_ID"
  )
  ) |> 
  select(DB_ID, GO_ID, Synonym)

# mapped_hits <- hits |> 
#   left_join(mapping) |> 
#   mutate(gene = str_remove(gene, "/.+")) |> 
#   arrange(gene) |> 
#   group_by(protein) |> 
#   slice_head(n = 1)

MF <- annotation |> 
  filter(namespace == "molecular_function") |> 
  select(GO_ID, "GO molecular function" = name)

BP <- annotation |> 
  filter(namespace == "biological_process") |> 
  select(GO_ID, "GO biological process" = name)

CC <- annotation |> 
  filter(namespace == "cellular_component") |> 
  select(GO_ID, "GO cellular component" = name)

# Collapse data frame to one row per blast hit

complete <- hits |> 
  left_join(gff) |> 
  left_join(terms, join_by("gene" == "DB_ID")) |> 
  left_join(MF) |> 
  left_join(BP) |> 
  left_join(CC) |> 
  group_by(gene) |> 
  mutate(across(where(is.character), ~ str_replace_na(., ""))) |> 
  summarise(
    gene = paste(unique(gene), collapse = "; "),
    Synonyms = paste(unique(Synonym), collapse = "|"),
    "GO terms" = paste(unique(GO_ID), collapse = "; "),
    `GO cellular component` = paste(unique(`GO cellular component`), collapse = "; "),
    `GO biological process` = paste(unique(`GO biological process`), collapse = "; "),
    `GO molecular function` = paste(unique(`GO molecular function`), collapse = "; "),
    Description = paste(unique(Description), collapse = "; ")
  ) |> 
  mutate(
    across(where(is.character), ~ str_remove_all(., "^; |; $")),
    across(where(is.character), ~ str_replace_all(., "; ; ", "; ")),
    across(where(is.character), ~ str_replace_all(., "^$", NA_character_)),
    Synonyms = str_split(Synonyms, fixed("|")),
    Synonyms = map(Synonyms, str_unique),
    Synonyms = map_chr(Synonyms, ~paste(., collapse = "; "))
    )

PKR_groups <- complete |> 
  summarise(
    "Total" = n(),
    "Kinase" = sum(str_detect(`GO molecular function`, "kinase") | str_detect(Description, "kinase"), na.rm = TRUE),
    "Phosphatase" = sum(str_detect(`GO molecular function`, "phosphatase") | str_detect(Description, "phosphatase"), na.rm = TRUE),
    "CW" = sum(str_detect(`GO cellular component`, "cell wall|extracellular|apoplast|surface"), na.rm = TRUE),
    "PM" = sum(str_detect(`GO cellular component`, "plasma membrane"), na.rm = TRUE)
    )

kable(PKR_groups)
```

## Simulation

```{r}
T1_selection <- tribble(
  ~Tray, ~`Seed weight [g]`, ~`Seed # min`, ~`Seed # max`, ~Transformants, ~`Efficiency [%]`,
  "1", 0.525, NA, NA, 36, NA,
  "2", 0.508, NA, NA, 93, NA,
  "3", 0.494, NA, NA, 56, NA,
  "4", 0.507, NA, NA, 57, NA,
  "5", 0.512, NA, NA, 44, NA,
  "6", 0.52, NA, NA, 113, NA,
  "7", 0.552, NA, NA, 111, NA,
  "8", 0.5, NA, NA, 81, NA,
  "9", 0.59, NA, NA, 83, NA,
  "10", 0.58, NA, NA, 73, NA
) |>
  mutate(`Seed # max` = round(`Seed weight [g]` / 0.00002, digits = 0),
         `Seed # min` = round(`Seed weight [g]` / 0.00003, digits = 0))

T1_total <- T1_selection |>
  drop_na(Transformants) |>
  summarise(across(where(is.numeric), ~sum(.x, na.rm = T))) |>
  mutate(Tray = "Screened")

screened <- T1_total |>  pull(Transformants)

T1_pred <- tribble(
  ~Tray, ~`Seed weight [g]`, ~`Seed count`, ~Transformants, ~`Efficiency [%]`,
  "Predicted", 32.5, NA, NA, NA,
) |>
  mutate(
    `Seed # max` = round(`Seed weight [g]` / 0.00002, digits = 0),
    `Seed # min` = round(`Seed weight [g]` / 0.00003, digits = 0),
    Transformants = round(T1_total$Transformants / T1_total$`Seed weight [g]` * `Seed weight [g]`, digits = 0)
  )

T1_selection <- T1_selection |>
  bind_rows(T1_total) |>
  bind_rows(T1_pred) |>
  unite("Seed count", `Seed # min`, `Seed # max`, sep = "–", remove = F) |>
  mutate(
    `Efficiency [%]` = case_when(
      !is.na(Transformants) ~ paste0(round(Transformants / `Seed # max` * 100, digits = 3), "–", round(Transformants / `Seed # min` * 100, digits = 3)),
      TRUE ~ ""
    )
  ) |>
  select(-c(`Seed # min`, `Seed # max`))

pred_T1 <- T1_pred$Transformants

sgRNAs <- 6161

unique_transformants <- tibble()
step <- 500
for (n in 1:50) {
    transformants <- tibble(transformants = c(1:22000)) |>
        mutate(sgRNA = sample(sgRNAs, n(), replace = TRUE))
    t_list <- as.list(transformants)
    for (i in c(seq(0, nrow(transformants), by = step), pred_T1, screened)) {
        y <- length(unique(t_list$sgRNA[t_list$transformants < (i + 1)]))
        unique_transformants <- rbind(unique_transformants, tibble(unique_sgRNAs = y, iteration = n, transformant = i))
    }
}

halfway <- unique_transformants |>
    group_by(iteration) |>
    filter(abs(unique_sgRNAs - (sgRNAs / 3)) == min(abs(unique_sgRNAs - (sgRNAs / 3)))) |>
    ungroup() |>
    summarise(
        range = paste0("<i>", min(transformant), "–", max(transformant), "</i> "),
        unique_sgRNAs = mean(unique_sgRNAs),
        transformant = mean(transformant)
    ) |>
    mutate(threshold = '33%')

nf <- unique_transformants |>
    group_by(iteration) |>
    filter(abs(unique_sgRNAs - (sgRNAs * 0.95)) == min(abs(unique_sgRNAs - (sgRNAs * 0.95)))) |>
    ungroup() |>
    summarise(
        range = paste0("<i>", min(transformant), "–", max(transformant), "</i> "),
        unique_sgRNAs = mean(unique_sgRNAs),
        transformant = mean(transformant)
    ) |>
    mutate(threshold = '<i>95%</i>')

current <- unique_transformants |>
  filter(transformant == screened) |>
  summarise(
        range = paste0("<i>", mean(transformant), "</i> "),
        unique_sgRNAs = mean(unique_sgRNAs),
        transformant = mean(transformant)
    ) |>
    mutate(threshold = paste0("<i>", round(unique_sgRNAs / sgRNAs * 100, digits = 0), "%</i>"))

pred <- unique_transformants |>
  filter(transformant == pred_T1) |>
  summarise(
        range = paste0("<i>", mean(transformant), "</i> "),
        unique_sgRNAs = mean(unique_sgRNAs),
        transformant = mean(transformant)
    ) |>
    mutate(threshold = paste0("<i>", round(unique_sgRNAs / sgRNAs * 100, digits = 0), "%</i>"))

thresholds <- bind_rows(current, pred, nf) |>
  arrange(desc(transformant))

sim_plot <- ggplot(unique_transformants, aes(x = transformant, y = unique_sgRNAs)) +
    geom_rect(
        data = thresholds,
        aes(
            xmax = transformant,
            xmin = -Inf,
            ymax = unique_sgRNAs,
            ymin = -Inf,
            fill = threshold
        )
    ) +
    geom_line(
        aes(group = iteration),
        alpha = 0.25,
        linewidth = 0.2
    ) +
    geom_richtext(
        data = thresholds,
        aes(
            x = transformant,
            y = unique_sgRNAs,
            label = threshold,
            colour = threshold
        ),
        hjust = 1,
        vjust = 0,
        family = "Futura",
        size = ggtext_size,
        label.size = NA,
        fill = NA,
        label.padding = unit(0, "lines"),
    ) +
    geom_richtext(
        data = thresholds,
        aes(
            x = transformant,
            y = unique_sgRNAs,
            label = range,
            colour = threshold
        ),
        hjust = 1,
        vjust = 1,
        angle = 90,
        family = "Futura",
        size = ggtext_size,
        label.size = NA,
        fill = NA,
        label.padding = unit(c(0.1, 0, 0, 0), "lines")
    ) +
    labs(
        y = 'Unique sgRNAs',
        x = 'Selected T1 lines') +
    scale_fill_manual(values = c("#d25952", "#e8c245", "#275d95"), guide = 'none') +
    scale_colour_manual(values = c("black", "white", "black"), guide = 'none') +
    scale_x_continuous(expand = expansion(add = 0)) +
    scale_y_continuous(expand = expansion(add = c(0, 200))) +
    coord_cartesian(xlim = c(0,19000), clip = "off") +
    theme_leo()

pdf(paste0(out_path, "PKR_sim.pdf"), width = onecol * 1.6, height = onecol*0.7)
sim_plot
dev.off()
```


## 2023-04 hypocotyls

```{r results='hide'}
onecm <- 236
data <- read_tsv(
  paste0(datapath, "2022_CRISPR_screen/phenotyping/2023-04-24_hypocotyls/lengths.tsv"),
  col_types = "fn"
) |>
  mutate(
    cm = px / 236
  )

letters <- letter_groups(
  data,
  cm,
  line,
  "kruskal",
  print_position = "above"
)

fills <- colorRampPalette(RColorBrewer::brewer.pal(11, "RdYlBu"))(12)

hyp_length <- ggplot(
  data,
  aes(
    x = reorder(line, -cm),
    y = cm
  )
) +
  geom_hline(
    yintercept = 1.6,
    linewidth = 0.2,
    linetype = 2
  ) +
  geom_violin(
    aes(fill = reorder(line, -cm)),
    draw_quantiles = 0.5,
    # fill = "white",
    colour = "black",
    alpha = 0.75,
    width = 0.5,
    size = 0.2,
    scale = "width"
  ) +
  geom_quasirandom(
    shape = 21,
    fill = "white",
    colour = "black",
    stroke = 0.1,
    alpha = 1,
    size = 1,
    width = 0.2
  ) +
  geom_text(
    data = letters,
    aes(label = Letters),
    size = ggtext_size,
    family = "IBMPlexMono"
  ) +
  labs(
    y = "Hypocotyl length [cm]",
    x = "T1 line #"
  ) +
  scale_fill_manual(values = rev(fills)) +
  theme_leo() +
  theme(legend.position = "none")

# pdf(paste0(out_path, "2023-04-24_hypocotyls.pdf"), width = onecol, height = onecol * 0.5)
hyp_length
# dev.off()

# print(hyp_length, vp = grid::viewport(gp = grid::gpar(cex = 2)))
```

## 2023-07 hypocotyls

```{r results='hide',fig.height=12,fig.width=8}
onecm <- 236
hyp_data <- read_tsv(
  paste0(datapath, "2022_CRISPR_screen/phenotyping/2023-07-10_hypocotyls/lengths.tsv"),
  col_types = "fnfn"
) |>
  mutate(
    cm = px / 236,
    line = ordered(line, levels = str_sort(levels(line), numeric = TRUE))
  )

letters <- letter_groups(
  hyp_data,
  cm,
  line,
  "kruskal",
  stat_alpha = 0.05,
  print_position = "above",
  print_adjust = 0.5
)

stars <- tibble()
for (i in 1:length(levels(hyp_data$line))) {
  pval <- with(hyp_data, t.test(cm[line == "Col-0"], cm[line == levels(hyp_data$line)[i]]))$p.value
  stars <- bind_rows(stars, tibble(line = levels(hyp_data$line)[i], pval = pval))
}
stars <- stars |>
  mutate(symbol = case_when(
    pval < 0.001 ~ "***",
    pval < 0.01 ~ "**",
    pval < 0.05 ~ "*",
    TRUE ~ NA
  ))

letters <- letters |>
  left_join(stars, by = "line")

fills <- colorRampPalette(RColorBrewer::brewer.pal(11, "RdYlBu"))(length(levels(hyp_data$line)))

sig_hyp <- hyp_data |>
  mutate(line = as.character(line)) |>
  group_by(line) |>
  summarise(median = median(cm)) |>
  filter(line %in% c("Col-0", (drop_na(stars, symbol) |> pull(line)))) |>
  mutate(sig_hyp = case_when(
    median < median[line == "Col-0"] ~ "↓",
    TRUE ~ "↑"
  )) |>
  filter(line != "Col-0") |>
  select(-median)


hyp_length <- ggplot(
  hyp_data,
  aes(
    x = reorder(line, -cm),
    # x = line,
    y = cm
  )
) +
  geom_violin(
    aes(fill = reorder(line, -cm)),
    draw_quantiles = 0.5,
    # fill = "white",
    colour = "black",
    alpha = 0.75,
    width = 0.5,
    linewidth = 0.2,
    scale = "width"
  ) +
  geom_quasirandom(
    aes(fill = reorder(line, -cm)),
    shape = 21,
    # fill = "white",
    colour = "black",
    stroke = 0.1,
    alpha = 1,
    size = 1,
    width = 0.2
  ) +
  geom_text(
    data = letters,
    aes(label = symbol),
    size = ggtext_size,
    #                 angle = 90,
    vjust = 0.75,
    hjust = 0,
    family = "Helvetica"
  ) +
  labs(
    y = "Hypocotyl length [cm]",
    x = "T1 line #"
  ) +
  scale_fill_manual(values = rev(fills)) +
  theme_leo() +
  theme(legend.position = "none") +
  coord_flip()

# pdf(paste0(out_path, "2023-07-10_hypocotyls.pdf"), height = twocol, width = onecol)
# hyp_length
# dev.off()
# 
# print(hyp_length, vp = grid::viewport(gp = grid::gpar(cex = 2)))
```


## 2023-08 roots

```{r results='hide',fig.height=12,fig.width=8}
onecm <- 236
root_data <- read_tsv(
  paste0(datapath, "2022_CRISPR_screen/phenotyping/2023-07-10_roots/lengths.tsv"),
  col_types = "fnfn"
) |>
  drop_na(px) |>
  mutate(
    cm = px / 236,
    line = ordered(line, levels = str_sort(levels(line), numeric = TRUE))
  )

letters <- letter_groups(
  root_data,
  cm,
  line,
  "kruskal",
  stat_alpha = 0.05,
  print_position = "above",
  print_adjust = 0.5
)

stars <- tibble()
for (i in 1:length(levels(root_data$line))) {
  pval <- with(root_data, t.test(cm[line == "Col-0"], cm[line == levels(root_data$line)[i]]))$p.value
  stars <- bind_rows(stars, tibble(line = levels(root_data$line)[i], pval = pval))
}
stars <- stars |>
  mutate(symbol = case_when(
    pval < 0.001 ~ "***",
    pval < 0.01 ~ "**",
    pval < 0.05 ~ "*",
    TRUE ~ NA
  ))

letters <- letters |>
  left_join(stars, by = "line")

fills <- colorRampPalette(RColorBrewer::brewer.pal(11, "RdYlBu"))(length(levels(root_data$line)))

sig_root <- root_data |>
  mutate(line = as.character(line)) |>
  group_by(line) |>
  summarise(median = median(cm)) |>
  filter(line %in% c("Col-0", (drop_na(stars, symbol) |> pull(line)))) |>
  mutate(sig_root = case_when(
    median < median[line == "Col-0"] ~ "↓",
    TRUE ~ "↑"
  )) |>
  filter(line != "Col-0") |>
  select(-median)

root_length <- ggplot(
  root_data,
  aes(
    x = reorder(line, -cm),
    # x = line,
    y = cm
  )
) +
  geom_violin(
    aes(fill = reorder(line, -cm)),
    draw_quantiles = 0.5,
    # fill = "white",
    colour = "black",
    alpha = 0.75,
    width = 0.5,
    linewidth = 0.2,
    scale = "width"
  ) +
  geom_quasirandom(
    aes(fill = reorder(line, -cm)),
    shape = 21,
    # fill = "white",
    colour = "black",
    stroke = 0.1,
    alpha = 1,
    size = 1,
    width = 0.2
  ) +
  geom_text(
    data = letters,
    aes(label = symbol),
    size = ggtext_size,
    #                 angle = 90,
    vjust = 0.75,
    hjust = 0,
    family = "Helvetica"
  ) +
  labs(
    y = "Root length [cm]",
    x = "T1 line #"
  ) +
  scale_fill_manual(values = rev(fills)) +
  theme_leo() +
  theme(legend.position = "none") +
  coord_flip()

# pdf(paste0(out_path, "2023-07-10_roots.pdf"), height = twocol, width = onecol)
# root_length
# dev.off()

# print(root_length, vp = grid::viewport(gp = grid::gpar(cex = 2)))
```

## Proposal figure


```{r}

highlights <- c("19", "153", "197", "454", "72", "666", "699", "720", "Col-0")

hyp_shadow <- hyp_data |> 
  group_by(line) |> 
  filter(!(line %in% highlights)) |> 
  summarise(count = n())

hyp_length_hl <- hyp_length +
  geom_tile(
    data = hyp_shadow,
    aes(x = line),
    y = 1,
    height = Inf,
    width = 0.7,
    fill = rgb(1, 1, 1, 0.8)
  ) +
  scale_x_discrete(breaks = highlights) +
  theme(axis.title.y = element_blank())

root_shadow <- root_data |> 
  group_by(line) |> 
  filter(!(line %in% highlights)) |> 
  summarise(count = n())

root_length_hl <- root_length +
  geom_tile(
    data = root_shadow,
    aes(x = line),
    y = 1,
    height = Inf,
    width = 0.7,
    fill = rgb(1, 1, 1, 0.8)
  ) +
  scale_x_discrete(breaks = highlights) +
  theme(axis.title.y = element_blank())

pdf(paste0(out_path, "UPSC_fig.pdf"), height = twocol*0.7, width = onecol*0.7)
root_length_hl
dev.off()

hyp_length + root_length
```

## Proposal Ganntt

```{r}
NNF <- tribble(
  ~Start, ~End, ~Aim, ~Task, ~FTE,
  1, 5, "1. PPI Networks", "Cloning & transformation", 0.4,
  5, 10, "1. PPI Networks", "Selection and sample preparation", 0.4,
  10, 12, "1. PPI Networks", "Proteomics and data analysis", 0.2,
  1, 5, "2. CRK39 & 40", "Cloning", 0.2,
  1, 5, "2. CRK39 & 40", "Allelic confirmation", 0.2,
  5, 9, "2. CRK39 & 40", "Phenotypic complementation", 0.2,
  9, 14, "2. CRK39 & 40", "Y2H/BiFC confirmation", 0.4,
  9, 12, "2. CRK39 & 40", "Phosphoproteomics", 0.2,
  9, 14, "2. CRK39 & 40", "Fluorescence microscopy", 0.4,
  12, 18, "2. CRK39 & 40", "Genetic conf. of interactors", 0.2,
  19, 24, "2. CRK39 & 40", "Manuscript preparation", 0.2,
  5, 22, "2. CRK39 & 40", "Transpose to poplar (Umeå)", 0.1,
  1, 24, "3. Screening for SCW regulators", "T1 screening", 0.2,
  3, 24, "3. Screening for SCW regulators", "Microscopy of T1 stems", 0.1,
  8, 24, "3. Screening for SCW regulators", "PUP-IT of candidates", 0.1
)

NNF_long <- NNF %>%
  # mutate(Start = isoweek(Start),
  #        End = isoweek(End)) %>%
  gather(date.type, task.date, -c(Aim, Task, FTE)) %>%
  arrange(date.type, task.date) %>%
  mutate(Task = factor(Task, levels=rev(unique(Task)), ordered=TRUE))

NNF_gantt <- ggplot(NNF_long, aes(x = forcats::fct_rev(forcats::fct_inorder(Task)), y = task.date, group = interaction(Aim, Task, FTE))) +
  geom_line(aes(
    linewidth = FTE,
    colour = Aim)
    ) +
  coord_flip(ylim = c(0, 24)) +
  labs(x = NULL,
       y = "Project month") +
  scale_colour_manual(values = pal_ostwald_disc) +
  scale_y_continuous(breaks = c(0, 6, 12, 18, 24)) +
  scale_linewidth(range = c(1, 5.5)) +
  theme_leo() +
  theme(axis.text.x = element_text(hjust = 0),
        panel.border = element_blank(),
        axis.ticks = element_blank(),
        legend.title = element_blank(),
        strip.text = element_text(face = "bold")) +
  ggforce::facet_col(~Aim, scales = "free_y", space = "free")

NNF_gantt

# pdf(paste0(out_path, "NNF_gantt_blaschek.pdf"), width = twocol, height = onecol)
# NNF_gantt
# dev.off()
```


## Select candidates

```{r}
lethal <- tibble(
  line = c("387", "462", "474", "633", "692"),
  Ster. = "🗸"
)

stem <- tibble(
  line = c("510", "657", "720"),
  Stem = "↓"
)

lines <- full_join(sig_root, sig_hyp) |>
  full_join(lethal) |>
  full_join(stem) |>
  separate(line, into = c("line", "T2 line"), sep = "-")

sgRNAs <- read_tsv("/home/leonard/Documents/Latex/2022_labbook/PKR_line_table.tsv") |>
  drop_na(Targets) |>
  mutate(Line = as.character(Line)) |>
  mutate(targets = str_extract_all(Targets, "[:alnum:]+")) |>
  select("line" = Line, targets, `Published phenotype`) |>
  unnest(targets)

write_tsv(sgRNAs, paste0(out_path, "phenotype_targets.tsv"))

hits <- sgRNAs |> 
  select("gene" = targets)

complete <- hits |> 
  left_join(gff) |> 
  left_join(terms, join_by("gene" == "DB_ID")) |> 
  left_join(MF) |> 
  left_join(BP) |> 
  left_join(CC) |> 
  group_by(gene) |> 
  mutate(across(where(is.character), ~ str_replace_na(., ""))) |> 
  summarise(
    gene = paste(unique(gene), collapse = "; "),
    Synonyms = paste(unique(Synonym), collapse = "|"),
    "GO terms" = paste(unique(GO_ID), collapse = "; "),
    `GO cellular component` = paste(unique(`GO cellular component`), collapse = "; "),
    `GO biological process` = paste(unique(`GO biological process`), collapse = "; "),
    `GO molecular function` = paste(unique(`GO molecular function`), collapse = "; "),
    Description = paste(unique(Description), collapse = "; ")
  ) |> 
  mutate(
    across(where(is.character), ~ str_remove_all(., "^; |; $")),
    across(where(is.character), ~ str_replace_all(., "; ; ", "; ")),
    across(where(is.character), ~ str_replace_all(., "^$", NA_character_)),
    Synonyms = str_split(Synonyms, fixed("|")),
    Synonyms = map(Synonyms, str_unique),
    Synonyms = map_chr(Synonyms, ~paste(., collapse = "; "))
    )

annotated_lines <- full_join(sgRNAs, complete, by = c("targets" = "gene")) |>
  full_join(lines)

write_tsv(annotated_lines, paste0(out_path, "phenotype_table.tsv"))
# filter(!(is.na(sig_root) & is.na(sig_hyp) & is.na(sterile))) |>
# filter(str_detect(`A. thaliana GO cellular component`, "membrane|apoplast|extracellular"))

table <- annotated_lines |>
  mutate(
    PM = case_when(
      str_detect(`GO cellular component`, "plasma membrane") ~ "🗸",
      TRUE ~ NA
    ),
    Wall = case_when(
      str_detect(`GO cellular component`, "apoplast|extracellular") ~ "🗸",
      TRUE ~ NA
    )
  ) |>
  mutate(line = ordered(as.factor(line), levels = str_sort(levels(as.factor(line)), numeric = TRUE))) |>
  # group_by(line) |>
  # summarise(across(everything(), ~ paste(unique(.x[!is.na(.x)]), collapse = "; "))) |>
  select(
    line,
    targets,
    `Published phenotype`,
    Description,
    "Root" = sig_root,
    "Hyp." = sig_hyp,
    Stem,
    Ster.,
    PM,
    Wall
  ) |>
  mutate(
    root_col = case_when(Root == "↑" ~ "#275d95", Root == "↓" ~ "#d25952", TRUE ~ "grey"),
    hyp_col = case_when(Hyp. == "↑" ~ "#275d95", Hyp. == "↓" ~ "#d25952", TRUE ~ "grey"),
    stem_col = case_when(Stem == "↑" ~ "#275d95", Stem == "↓" ~ "#d25952", TRUE ~ "grey"),
    sterile_col = case_when(Ster. == "🗸" ~ "#e8c245", TRUE ~ "grey"),
    pm_col = case_when(PM == "🗸" ~ "#e8c245", TRUE ~ "grey"),
    wall_col = case_when(Wall == "🗸" ~ "#e8c245", TRUE ~ "grey")
  )


kable(
  table |> select(-c(11:16)) |> drop_na(targets),
  "html",
  caption = "Summary of identified CRISPR targets.",
  escape = F,
  align = "llllcccccc",
) |>
  column_spec(5, background = table$root_col) |>
  column_spec(6, background = table$hyp_col) |>
  column_spec(7, background = table$stem_col) |>
  column_spec(8, background = table$sterile_col) |>
  column_spec(9, background = table$pm_col) |>
  column_spec(10, background = table$wall_col) |>
  column_spec(2:4, border_left = "10px solid white") |>
  column_spec(5:10, border_left = "2px solid white") |>
  row_spec(1:nrow(table |> drop_na(targets)), extra_css = "border-bottom: 2px solid white")
```

## 2024 table

### data

```{r}
total <- table |> 
  distinct(line, .keep_all = TRUE) |> 
  replace_na(list(Ster. = "", Stem = "", Hyp. = "", Root = "")) |> 
  summarise(
    `Abberant phenotype (T1)` = n(),
    Sterile = sum(Ster. != ""),
    Stem = sum(Stem != ""),
    Hypocotyl = sum(Hyp. != ""),
    Root = sum(Root != ""),
  ) |> 
  mutate(group = "Total", .before = 1)

seq <- table |> 
  distinct(line, .keep_all = TRUE) |> 
  replace_na(list(Ster. = "", Stem = "", Hyp. = "", Root = "")) |> 
  filter(targets != "") |> 
  summarise(
    `Abberant phenotype (T1)` = n(),
    Sterile = sum(Ster. != ""),
    Stem = sum(Stem != ""),
    Hypocotyl = sum(Hyp. != ""),
    Root= sum(Root != ""),
  )|> 
  mutate(group = "Sequenced", .before = 1)

new <- table |> 
  distinct(line, .keep_all = TRUE) |> 
  replace_na(list(Ster. = "", Stem = "", Hyp. = "", Root = "")) |> 
  filter(`Published phenotype` == "none") |> 
  summarise(
    `Abberant phenotype (T1)` = n(),
    Sterile = sum(Ster. != ""),
    Stem = sum(Stem != ""),
    Hypocotyl = sum(Hyp. != ""),
    Root= sum(Root != ""),
  ) |> 
  mutate(group = "Completely unknown function", .before = 1)

published <- table |> 
  distinct(line, .keep_all = TRUE) |> 
  replace_na(list(Ster. = "", Stem = "", Hyp. = "", Root = "")) |> 
  filter(`Published phenotype` != "none" & `Published phenotype` != "") |> 
  summarise(
    `Abberant phenotype (T1)` = n(),
    Sterile = sum(Ster. != ""),
    Stem = sum(Stem != ""),
    Hypocotyl = sum(Hyp. != ""),
    Root = sum(Root != ""),
  ) |> 
  mutate(group = "Partial phenotype published", .before = 1)

sum_tab <- bind_rows(total, seq, new, published)
```

### table

```{r}
sum_gt <- sum_tab |> 
  gt(rowname_col = "group") |> 
  tab_spanner("Abberant phenotype (T2)", columns = 3:6) |> 
  tab_stub_indent(2, indent = 2) |> 
  tab_stub_indent(3:4, indent = 4) |> 
  tab_caption("Table 1: Preliminary screening yielded 747 transformants, of which 62 lines had severe T1 phenotypes.")

sum_gt 
  # gtsave("table1.html")
```


